File: java/src/main/java/com/gpayments/requestor/testlab/config/RestClientConfig.java

/*
 *  Copyright (C) GPayments Pty Ltd - All Rights Reserved
 *  Copying of this file, via any medium, is subject to the
 *  ActiveServer End User License Agreement (EULA)
 *
 *  Proprietary code for use in conjunction with GPayments products only
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Written by GPayments <techsupport@gpayments.com>, 2019
 *
 *
 */

package com.gpayments.requestor.testlab.config;

import java.io.File;
import java.io.IOException;
import java.security.KeyManagementException;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.UnrecoverableKeyException;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.util.Collection;
import javax.net.ssl.SSLContext;
import org.apache.http.conn.ssl.TrustSelfSignedStrategy;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.ssl.SSLContextBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.web.client.RestTemplate;

@Configuration
public class RestClientConfig {

  private static final String KEYSTORE_PASSWORD = "123456";
  private static final String KEY_ENTRY_PASSWORD = "123456";
  private static final String CA_CERTS_FILE_NAME = "certs/cacerts.pem";

  @Bean
  public RestTemplate restTemplate(Config config)
      throws IOException, UnrecoverableKeyException, CertificateException, NoSuchAlgorithmException,
      KeyStoreException, KeyManagementException {

    SSLContext sslContext =
        SSLContextBuilder.create()
            .loadKeyMaterial(
                new File(config.getCertFileName()),
                KEYSTORE_PASSWORD.toCharArray(),
                KEY_ENTRY_PASSWORD.toCharArray())
            .loadTrustMaterial(getCAKeystore(CA_CERTS_FILE_NAME), new TrustSelfSignedStrategy())
            .build();

    CloseableHttpClient client = HttpClients.custom().setSSLContext(sslContext).build();
    HttpComponentsClientHttpRequestFactory httpRequestFactory =
        new HttpComponentsClientHttpRequestFactory(client);

    return new RestTemplate(httpRequestFactory);
  }

  private KeyStore getCAKeystore(String fileName)
      throws CertificateException, IOException, KeyStoreException, NoSuchAlgorithmException {

    CertificateFactory factory = CertificateFactory.getInstance("X.509");

    Collection<? extends Certificate> certificates =
        factory.generateCertificates(new ClassPathResource(fileName).getInputStream());

    KeyStore keystore = KeyStore.getInstance("JKS");
    keystore.load(null);

    int id = 0;
    for (Certificate cert : certificates) {
      keystore.setCertificateEntry("cert-alias-" + (++id), cert);
    }

    return keystore;
  }
}

File: php/config/RestClientConfig.php
<?php session_start();

use GuzzleHttp\Exception\BadResponseException;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\MessageFormatter;
use GuzzleHttp\Middleware;
use Monolog\Handler\StreamHandler;
use Monolog\Logger;

/**
 *  Copyright (C) GPayments Pty Ltd - All Rights Reserved
 *  Copying of this file, via any medium, is subject to the
 *  ActiveServer End User License Agreement (EULA)
 *
 *  Proprietary code for use in conjunction with GPayments products only
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Written by GPayments <techsupport@gpayments.com>, 2020
 *
 *
 */
class RestClientConfig
{

    const CA_CERTS_FILE_PATH = "resources/certs/cacerts.pem";

    private $client;
    private $headers = [];

    public function __construct(Config $config)
    {
        $logger = new Logger('Logger');
        $logger->pushHandler(new StreamHandler('php://stdout', Logger::DEBUG)); // <<< uses a stream
        $stack = HandlerStack::create();
        $stack->push(
            Middleware::log(
                $logger,
                new MessageFormatter('{req_body} - {res_body}')
            )
        );
        $this->client = new GuzzleHttp\Client([
            'base_uri' => $config->getAsAuthUrl(),
            'cert' => [$config->getCertFileName(), ''],
            'ssl_key' => $config->getKeyFileName(),
            "verify" => self::CA_CERTS_FILE_PATH,
            'handler' => $stack,
        ]);

        if ($config->isGroupAuth()) {
            $this->headers = array('AS-Merchant-Token' => $config->getMerchantToken());
        }
    }

    function post($request_uri, $post_data)
    {
        try {
            return $this->client->request(
                "POST",
                $request_uri,
                ['json' => $post_data, 'headers' => $this->headers]);

        } catch (BadResponseException $e) {
            Utils::_returnJson($e->getResponse()->getBody(), $e->getCode());
        } catch (GuzzleException $e) {
            Utils::_returnJson($e->getMessage(), 500);
        }
    }

    function get($request_uri)
    {
        try {
            return $this->client->request("GET", $request_uri, ['headers' => $this->headers]);
        } catch (BadResponseException $e) {
            Utils::_returnJson($e->getResponse()->getBody(), $e->getCode());
        } catch (GuzzleException $e) {
            Utils::_returnJson($e->getMessage(), 500);
        }
    }

}


